### DISCUSSION TITLE
How to make only some props optional

### DISCUSSION THREAD

**Discussion by psi-4ward** [sinclairzx81/typebox#1125] [Q&A] [answered]
I want to derive a new Schema where some props are optional and came up with the following solution (drizzle-typebox):

```ts
import {createSelectSchema} from 'drizzle-typebox';

export const filesTbl = table(
  'files',
  {
    id: integer().primaryKey({autoIncrement: true}),
    ...tstamps,
    name: text({length: 255}).notNull(),
    contentType: text({length: 255}).notNull().default('application/octet-stream'),
    size: integer().notNull()
  },
);

export const filesSelectSchema = createSelectSchema(filesTbl);
export const filesInsertSchema = Type.Object({
  ...Type.Omit(filesSelectSchema, ['id', 'created_at', 'updated_at', 'contentType']).properties,
  contentType: Type.Optional(filesSelectSchema.properties.contentType), // We have a default value for contentType
});
```

So I thought about how this could be expressed in a more intuitive way like `Type.Optional(fileSelectSchema, ['contentType'])` so we could give a list of props which should be optional.

Is there a better way to do this? What do you think about extending the Optional() method?

Greets

**Comment by sinclairzx81** [comment, marked as answer]
@psi-4ward Hi, Try the following

[TypeScript Link Here](https://www.typescriptlang.org/play/?moduleResolution=99&module=199#code/JYWwDg9gTgLgBAbzgFQJ5gKZwL5wGZQQhwDkAAgM7AB2AxgDYCGwUA9DOhgEYQAeJAKAGtWcALQTJU6TNlz5C+cNFpMAZVpRgYGMvGKDho4aEdMcAMJFIVGBgA8yAHxwAvIgDaAaTg04AawxUCDwUAF0ALhRvMNwAMkRcETgoDDB6YFpGOxQAIThaawhbDFNOOAAxYHoMNQwa2nh3BAE4XwATKIoYLWoAcwAaVoLU7Ix2gH1sqOoAVxAuDChh2bB2scnpuDmFpeHC6jtD1Qwunpo+4bxqjGpGEFO4bt7LtvaMCk1tGGAIajOXgJsGVzFUagBJagUJZNSxFEr2ADyIGAMHsYNq9QwjQGpGA7RIcAAPqRNBgNlMYISSSRVus7JsqcTSX8jjATiQXHFhm1eXz+W0AAqMWDARj0eyCzL+dE3OoNGC4kgHNkcpzqoRCZLGHW6nV6E65Ph6PWms2yAQHbqVG4UeXY2EnAB0iK4ACsHQAKFptfFRZ1qc79T0ASiGbTJFK2zoAcvNFlBQ+G4HSozB-ZwnXHdomw-tWbd2ZwM5gnYGXkmrjc7g8Sxgy0G+pW3h8vjpfv8UJnyxdQ0CQ0IrfAMRRIdDYG4u6WrOBiqiMJ6PMNncjUZ6R-acXAPCR8SQlZGGZT96RU0fsiflQXjpwSGE821ncLReLPU-pevbZvFdur4dCxy955vempAA)

```typescript
import { Type } from '@sinclair/typebox'

// ----------------------------------------------------------------
// TypeScript
// ----------------------------------------------------------------

type Composite<T> = {[K in keyof T]: T[K]} & {} // replicate TB composite

type FileSelect = {
  id: string,
  created_at: number
  updated_at: number
  contentType: string
  filename: string
  description: string
}

type FileInsert = Composite<
  Omit<FileSelect, 'id' | 'created_at' | 'updated_at' | 'contentType'> &
  Partial<Pick<FileSelect, 'contentType'>>
>



// ----------------------------------------------------------------
// TypeBox
// ----------------------------------------------------------------
const FilesSelect = Type.Object({
  id: Type.String(),
  created_at: Type.Number(),
  updated_at: Type.Number(),
  contentType: Type.String(),
  filename: Type.String(),
  description: Type.String()
})

const FilesInsert = Type.Composite([
  Type.Omit(FilesSelect, ['id', 'created_at', 'updated_at', 'contentType']),
  Type.Partial(Type.Pick(FilesSelect, ['contentType'])),
])
```

It is generally recommended to construct types using the TB API where possible. TypeBox is mostly aligned to how you would construct the same type using TypeScript. While it is possible construct properties using the `...` spread operator, working through the API generally leads to types more closely aligned to their TypeScript counterparts.

Hope this helps!
S

**Reply by psi-4ward** [reply]
Very nice! Thank you!!

```ts
function deriveInsertSchema<T extends TSchema>(t: T) {
  return Type.Composite([
    Type.Omit(t, ['id', 'createdAt', 'updatedAt', 'contentType']),
    Type.Partial(Type.Pick(t, ['contentType'])),
    Type.Optional(Type.Object({updatedAt: Type.Number({minimum: 0})}))
  ])
}
```

This is an example of a simple helper functions to do the modifications and looks like it works!

**Comment by DIEGOHORVATTI** [comment]
Very interesting, it could work at any level of the hierarchy, the name for the example, omitting the required from any level of the schema hierarchy would be very useful today, do this with a recursive function
